core/vm/analysis.go

```diff
@@ -116,3 +116,86 @@ func codeBitmapInternal(code, bits bitvec) bitvec {
 	}
 	return bits
 }
+
+// eofCodeBitmap collects data locations in code.
+func eofCodeBitmap(code []byte) bitvec {
+	// The bitmap is 4 bytes longer than necessary, in case the code
+	// ends with a PUSH32, the algorithm will push zeroes onto the
+	// bitvector outside the bounds of the actual code.
+	bits := make(bitvec, len(code)/8+1+4)
+	return eofCodeBitmapInternal(code, bits)
+}
+
+// eofCodeBitmapInternal is the internal implementation of codeBitmap for EOF
+// code validation.
+func eofCodeBitmapInternal(code, bits bitvec) bitvec {
+	for pc := uint64(0); pc < uint64(len(code)); {
+		op := OpCode(code[pc])
+		pc++
+
+		// RJUMP and RJUMPI always have 2 byte operand.
+		if int8(op) == int8(RJUMP) || int8(op) == int8(RJUMPI) {
+			bits.setN(set2BitsMask, pc)
+			pc += 2
+			continue
+		}
+		var numbits uint8
+		if int8(op) >= int8(PUSH1) && int8(op) <= int8(PUSH32) {
+			numbits = uint8(op - PUSH1 + 1)
+		} else if int8(op) == int8(RJUMPV) {
```

Comment:
For one, I'm not sure why RJUMP and RJUMPI have special treatment? IMO PUSH should be checked first, because those are probably more common. 

e.g. something like this 
```
switch{
  case op >= PUSH1 && op <= PUSH32:
     numbits = uint8(op - PUSH1 + 1)
  case op == RJUMP || op == RJUMPI:
    numbits =2
  case op == RJUMPV:
   ... 
   default: 
    continue
}
```

}

Reply:
And secondly, comparisons on equality doesn't require casts. So `int8(op) == int8(RJUMPV)` should be written as `op == RJUMPV`. 
I suppose you might need the cast when doing less-than and greater-than comparisons (?)